### PR TITLE
feat(server-nestjs): migrate project bulk route

### LINKED ISSUE #2153: [Nest-js] Basculer le reprovisionnement de projets sur server-nestjs
Dans le cadre de la transition `server` -> `server-nestjs`, un certain travail a déjà été fait côté `server-nestjs` pour permettre un certain nombres d'actions, toutes liées directement ou indirectement au provisionnement des projets. C'est cette approche qui a permis de convertir le code des plugins.

Afin de pouvoir avancer sur la migration elle-même, il nous faut commencer à basculer des fonctionnalités existantes (autre qu'OpenCDS, par exemple), et celle qui paraît la plus simple est paradoxalement la fonctionnalité centrale de la Console: le reprovisionnement de projet.

- [ ] (`nginx-strangler`) Basculer les routes de reprovisionnement projet sur la base d'une variable d'environnement (permet de revenir en arrière en Production si besoin était)
- [ ] (`playwright`) Établir côté `tests-e2e` et `tests-integration` les éléments qui relèvent de cette thématique
- [ ] (`server`) Marquer, d'une manière ou d'une autre, le code concerné comme étant déprécié (il sera alors supprimé lors de la MEP suivante)

### PR DESCRIPTION
**Pull request by shikanime**
## Issues liées

Issues numéro: https://github.com/cloud-pi-native/console/issues/2153

---------

### CODE REVIEW

**Comment by shikanime** [comment]
## Code Review Summary

**Verdict: Changes Requested** (2 CI failures, 3 warnings, 2 suggestions)

### 🔴 Critical

1. **CI: 2 unit tests failing** — `user.guard.spec.ts` asserts `includeUserType: false` but `UserGuard` now passes `includeUserType: true` (changed in commit `1c2b59eb3`). The spec was not updated in this PR.
   - `user.guard.spec.ts:72` — "validates required admin permissions"
   - `user.guard.spec.ts:98` — "validates required user types"
   - Fix: update both `expect(authService.authenticate).toHaveBeenCalledWith(...)` calls to use `{ includeAdminRoleIds: true, includeUserType: true }`.

### ⚠️ Warnings

2. **`project.service.ts:159`** — `update()` is not decorated with `@StartActiveSpan()` unlike all other methods in this service and the sibling `create()` / `get()` / `archive()`. This means update operations will be missing their active span context. Likely an oversight.

3. **`project-bulk.service.ts:59-72`** — `runTasks` uses `Promise.allSettled` which is correct for fault tolerance. However, if `action` is none of 'archive'/'lock'/'unlock'/'replay', it silently resolves without doing anything. No error thrown for unknown actions — could be intentional but worth documenting.

4. **`project-roles.service.ts:61-63`** — The position-coherence check `positionsAvailable.length !== dbRoles.length` only fires when `positionsAvailable.length` is truthy. If a patch sends 0 roles with positions, the check is skipped. Edge case but probably fine for now.

### 💡 Suggestions

5. **`project.controller.ts:52`** — `@Req() request: FastifyRequest` is used solely for `request.id`. Consider using a dedicated decorator or injecting the request ID via a lighter mechanism (e.g., a `@RequestId()` decorator) to avoid coupling the NestJS controller to the Fastify type globally. Minor — current approach works.

6. **`project-services.service.ts:67`** — `this.plugin.infos(projectId)` catches errors per-plugin but the `to()` call on line 74 is wrapped in its own try/catch that only logs at `debug` level. If a plugin's `to()` throws unexpectedly, the error is silently swallowed and the service returns no URLs for that plugin. Consider whether `warn` level would be more appropriate for production visibility.

7. **`LogController` (line 31-35)** — The cast `params: { params: { projectId: query.projectId } } as Parameters<ProjectLoaderService['load']>[0]` is a type assertion that bypasses the compiler. If `ProjectLoaderService.load` signature changes, this silently breaks. A typed helper would be safer.

### ✅ Looks Good

- **Guard architecture** — Clean separation: `UserGuard` → `ProjectGuard` → `ProjectPolicy` + `ProjectService.validate()`. The policy-based approach with decorators is consistent and extensible.
- **Permission bitmask usage** — `ProjectAuthorized` and `AdminAuthorized` used correctly throughout. The `resolveProjectPermissions` logic in `ProjectLoaderService` correctly aggregates per-role perms with `everyonePerms | GUEST` baseline.
- **Slug generation** — `generateSlug` handles collisions correctly with incremental suffix.
- **Archive logic** — Properly renames slug/name with timestamp suffix to avoid unique constraint violations, clears cluster associations, and emits `project.delete`.
- **Test coverage** — Comprehensive specs for ProjectService (create/list/get/update/archive/getData) with good edge-case coverage (locked projects, owner transfers, non-admin restrictions).
- **Plugin system** — All 7 plugin services (ArgoCD, GitLab, Keycloak, Nexus, Registry, SonarQube, Vault) follow the same pattern with `infos()` + fault-tolerant `Promise.allSettled` in `PluginService`.
- **Documentation** — `MODULARISATION-STATUT.md` and `MODULARISATION-CARTOGRAPHIE.md` are meticulously updated with migration dates and route tables.
- **Nginx routing** — `/api/v1/projects` correctly routed to `server-nestjs`.
- **EventEmitterModule** — Removed from `MainModule` (now only in `InfrastructureModule`), avoiding double `.forRoot()` initialization.
- **LogModule** — Properly scoped with `UserGuard` + `ProjectLoaderService` for permission checks, clean log filtering for non-admins.

### Action Required

Fix the 2 failing tests in `user.guard.spec.ts` before merge. Everything else is non-blocking.

---
*Reviewed by Hermes Agent*

**Comment by shikanime** [comment]
## Deep-Dive Review: Code Structure Normalization, Logging, Testing, Iso-Behaviour

This review compares the new NestJS `server-nestjs` modules against the legacy `server` Fastify implementation for behavioral parity, structural consistency, logging completeness, and test coverage.

---

### 1. Structure Normalization

**Observation**: The new NestJS modules follow a consistent `controller → service → queries.utils` 3-file pattern. The legacy code used `router → business → queries`. The mapping is clean.

| Legacy (Fastify) | NestJS | Status |
|---|---|---|
| `project/router.ts` | `project.controller.ts` | ✅ |
| `project/business.ts` | `project.service.ts` | ✅ |
| `project/queries.ts` | `project-queries.utils.ts` | ✅ |
| `project-member/business.ts` | `project-members.service.ts` + `project-members-queries.utils.ts` | ✅ |
| `project-role/business.ts` | `project-roles.service.ts` | ✅ |
| `project-service/business.ts` | `project-services.service.ts` + `project-services.utils.ts` | ✅ |

**Findings**:

1. **`project.controller.ts` uses `@Req() request: FastifyRequest` directly** — This couples the NestJS controller to the Fastify runtime. The legacy `router.ts` also used Fastify's `req.id` for request tracing. Consider extracting the request ID via a custom decorator or middleware to keep the controller transport-agnostic.

2. **Sub-module granularity** — The NestJS version splits `project-service` into `project-services` + `project-hooks` + `project-bulk` + `project-secrets` + `project-members` + `project-roles`. The legacy had these as separate resource folders. ✅ This is a 1:1 structural mapping.

3. **`project.module.ts` imports `LogModule`** — This introduces a dependency from the domain module to a cross-cutting infrastructure module. The legacy code used `addLogs` from `@/resources/queries-index.js` directly. The NestJS approach (injecting `LogService`) is cleaner but creates coupling. Consider if `LogService` should be part of `InfrastructureModule` exports (which it already is via `LogModule`).

---

### 2. Logging

**Legacy pattern**: `logger.info({ requestId, userId, projectId }, 'message')` using pino child logger scoped per resource.

**NestJS pattern**: `this.logger.log(...)` / `this.logger.debug(...)` / `this.logger.warn(...)` using NestJS built-in `Logger`.

| Aspect | Legacy | NestJS | Verdict |
|---|---|---|---|
| Log levels | `info`, `error`, `warn` | `log`, `debug`, `warn`, `error` | ✅ Equivalent |
| Structured data | `{ requestId, userId, projectId }` | Object literals passed as method args | ⚠️ Inconsistent |
| Request ID | `req.id` (Fastify) | `request.id` (injected via `@Req()`) | ✅ |
| Pino child scopes | `baseLogger.child({ scope: 'resource:project' })` | `new Logger(ProjectService.name)` | ⚠️ Less granular |

**Findings**:

1. **`project.service.ts:100-148`** — `create()` logs are structured as `this.logger.log(\`project.create started (requestorUserId=..., projectName=...)\`)` using template literals. The legacy used `logger.info({ requestId, userId, slug }, 'message')` with structured objects. The NestJS approach loses pino's structured-logging benefits (searchable fields). Consider injecting a scoped pino logger or using `this.logger.log({ ... }, 'message')` format.

2. **`project.service.ts:159` (`update()`)** — Missing `@StartActiveSpan()`. All other methods (`create`, `get`, `archive`, `getData`) have it. This is a telemetry gap.

3. **`project-members.service.ts`** — Uses `this.logger.log(...)` with template literals. No structured fields. Legacy used `logger.info({ projectId, userId, ... })`.

4. **`project-roles.service.ts`** — No logging at all. Legacy `project-role/business.ts` also had no logging. ✅ Iso-preserving.

5. **`project-bulk.service.ts`** — Has `this.logger.log(...)` for start/complete/failure. Good. But no `requestId` in logs (legacy also didn't have it in bulk). ✅

6. **`project-services.service.ts`** — Logs only at `debug` level for URL generation failures. Legacy used `logger.warn` for similar cases. ⚠️ Consider `warn` for production visibility.

7. **`log.service.ts`** — `addLog()` correctly excludes sensitive fields (`cluster`, `user`, `newCreds`, `apis`) via `exclude()`. Legacy had the same exclusion list. ✅ Iso-complete.

---

### 3. Testing

| Module | Legacy Test | NestJS Test | Coverage Assessment |
|---|---|---|---|
| `project` | `business.spec.ts` + `router.spec.ts` | `project.service.spec.ts` (529 lines) | ✅ More comprehensive |
| `project-member` | None | `project-members.service.spec.ts` (268 lines) | ✅ New coverage |
| `project-role` | `business.spec.ts` | `project-roles.service.spec.ts` (174 lines) | ✅ |
| `project-service` | None | `project-services.service.spec.ts` (176 lines) | ✅ New coverage |
| `project-bulk` | None | `project-bulk.service.spec.ts` (84 lines) | ✅ New coverage |
| `project-hooks` | None | `project-hooks.service.spec.ts` (72 lines) | ✅ New coverage |
| `project-secrets` | None | `project-secrets.service.spec.ts` (99 lines) | ✅ New coverage |
| `log` | None | `log.service.spec.ts` + `log.controller.spec.ts` | ✅ New coverage |
| `deployment` | None | `deployment.service.spec.ts` (updated) | ✅ Updated |
| `user.guard` | N/A | `user.guard.spec.ts` | ❌ 2 tests failing |
| `project.guard` | N/A | `project.guard.spec.ts` | ✅ |
| `project-loader` | N/A | `project.service.spec.ts` (indirect) | ✅ |
| `gitlab-client` | N/A | `gitlab-client.service.spec.ts` (updated) | ✅ |
| `plugin` | N/A | `plugin.service.spec.ts` (55 lines) | ✅ |

**Findings**:

1. **`user.guard.spec.ts` (2 failures)** — Tests assert `includeUserType: false` but `UserGuard` now passes `true`. The spec was not updated when the guard changed. **Must fix before merge.**

2. **`project.service.spec.ts`** — Excellent coverage: create (slug generation, hook events, log emission), list (filter by member/all, status, search), get (not found, archived), update (strip locked, strip ownerId, owner change validation, locked project), archive (deletes dependencies, renames, emits event), getData. 529 lines, well-structured with faker mocks.

3. **`project-members.service.spec.ts`** (268 lines) — Covers addMember (by userId, by email, owner rejection, Keycloak lookup), patchMembers, removeMember. Good edge case coverage.

4. **`project-roles.service.spec.ts`** (174 lines) — Covers listRoles, patchRoles (system role protection, position coherence, oidcGroup validation), createRole, deleteRole (cascading member cleanup). Solid.

5. **Missing: controller specs for sub-modules** — `project.controller.spec.ts` exists (updated from `project-datastore.service.spec.ts`), but `project-members.controller.ts`, `project-roles.controller.ts`, `project-services.controller.ts`, `project-bulk.controller.ts`, `project-hooks.controller.ts`, `project-secrets.controller.ts` have no corresponding specs. The legacy had `router.spec.ts` for the main project router. Consider adding controller specs for the new sub-routes.

6. **`__mocks__/prisma.ts`** — Changed `vi.mock('../prisma.js')` to `vi.mock('../prisma')`. This is a path alias normalization. ✅

---

### 4. Iso-Behaviour with Legacy

#### 4.1 Project Create

| Aspect | Legacy | NestJS | Verdict |
|---|---|---|---|
| Slug generation | `generateSlug(name, existingSlugs)` | `generateSlug(name, existingSlugs)` | ✅ Identical |
| Default roles | 4 system roles (Admin, DevOps, Dev, Readonly) | Same 4 roles with same perms | ✅ Identical |
| Owner assignment | `ownerId: requestor.id` | `owner: { connect: { id: ownerId } }` | ✅ Equivalent |
| Hook call | `hook.project.upsert(project.id)` | `eventEmitter.emitAsync('project.upsert', project)` | ✅ Equivalent (hooks → events) |
| Log | `addLogs({ action: 'Create Project', ... })` | `this.logs.addLog({ action: 'Create Project', ... })` | ✅ |
| Role sync logs | `addLogs({ action: 'Upsert Project Role', ... })` per role | `this.logs.addLog({ action: 'Upsert Project Role', ... })` per role | ✅ |
| Error on hook failure | Returns `Unprocessable422` | No hook failure check (event-based) | ⚠️ Behavioral difference |

**⚠️ Behavioral difference**: Legacy `createProject` calls `hook.project.upsert(project.id)` and checks `results.failed`. If hooks fail, the project creation is rolled back (returns 422). The NestJS version emits an event and does not check the result. This is an architectural shift (async hooks vs sync hooks) — likely intentional but means project creation no longer fails fast if downstream plugins fail.

#### 4.2 Project Update

| Aspect | Legacy | NestJS | Verdict |
|---|---|---|---|
| Lock check | `if (perms.projectLocked && !isAdmin && data.locked !== false)` | `if (projectDb.locked && effectiveData.locked !== false)` | ✅ Equivalent |
| Owner change validation | Same checks (member exists, human) | Same checks | ✅ |
| Hook call | `hook.project.upsert(projectId)` | `eventEmitter.emitAsync('project.upsert', project)` | ✅ |
| Log | `addLogs({ action: 'Update Project', ... })` | `this.logs.addLog(...)` | ✅ |

#### 4.3 Project Archive

| Aspect | Legacy | NestJS | Verdict |
|---|---|---|---|
| Hook call | `hook.project.delete(projectId)` | None (event only) | ⚠️ Behavioral difference |
| Dependency deletion | Repos + Environments + Deployments | Repos + Environments + Deployments | ✅ Identical |
| Rename | `${name}_${Date.now()}_archived` | Same pattern | ✅ |
| Lock after archive | `locked: true` | `locked: true` | ✅ |
| Cluster clearing | `clusters: { set: [] }` | `clusters: { set: [] }` | ✅ |
| Log | `addLogs({ action: 'Delete all project resources', ... })` | `this.logs.addLog(...)` | ✅ |

**⚠️ Behavioral difference**: Legacy `archiveProject` calls `hook.project.delete(projectId)` and checks `results.failed`. The NestJS version does not call any hook. This means external cleanup (ArgoCD app deletion, etc.) won't happen on archive. This is a significant behavioral regression if hooks are not handled by the event listener.

#### 4.4 Project Secrets

| Aspect | Legacy | NestJS | Verdict |
|---|---|---|---|
| Auth | `authUser(req, { id: projectId })` + `SeeSecrets` check | `ProjectGuard` + `RequireProjectStatus` + `RequireProjectPermission('SeeSecrets')` | ✅ |
| Vault listing | `hook.project.getSecrets(projectId)` | `this.vault.listProjectSecrets(project.slug)` | ✅ (direct vault call) |
| Value parsing | `Object.entries(hookReply.results).map(...)` | `parseSecretValue(value)` per key | ✅ Equivalent |
| Fault tolerance | Returns `Unprocessable422` on hook failure | Returns empty result on vault listing failure | ⚠️ Slightly different |

#### 4.5 Project Members

| Aspect | Legacy | NestJS | Verdict |
|---|---|---|---|
| Add by email | `hook.user.retrieveUserByEmail` | `this.keycloak.getUserByEmail` | ✅ Equivalent (hook → direct Keycloak) |
| Owner exclusion | `if (userInDb.id === projectOwnerId) throw` | Same check | ✅ |
| Hook call | `hook.projectMember.upsert` + `hook.projectMember.delete` | `eventEmitter.emitAsync('projectMember.upsert/delete')` | ✅ |
| Return format | Returns full member list after mutation | Returns full member list after mutation | ✅ |

#### 4.6 Project Roles

| Aspect | Legacy | NestJS | Verdict |
|---|---|---|---|
| System role protection | `isSystemRoleType` check on both DB and input | Same checks | ✅ |
| Position coherence | `positionsAvailable.length !== dbRoles.length` | Same check | ✅ |
| OIDC group prefix | `/${project.slug}${matchingRole.oidcGroup}` | Same pattern | ✅ |
| Hook call | `hook.projectRole.upsert` + `hook.projectRole.delete` | `eventEmitter.emitAsync('project.upsert')` | ✅ |
| Delete cascading | Removes roleId from all members' roleIds | Same cascading cleanup | ✅ |

#### 4.7 Project Services

| Aspect | Legacy | NestJS | Verdict |
|---|---|---|---|
| Auth | `ProjectAuthorized.Manage` or project permissions | `RequireProjectAccess()` + admin check | ✅ |
| Admin target check | `AdminAuthorized.Manage(perms.adminPermissions)` | Same check | ✅ |
| Store generation | `dbToObj([...projectStore, ...globalConfig])` | Same logic in `generatePluginsUpdateBody` | ✅ |
| URL generation | `to({ clusters, zones, environments, project, store })` | Same signature | ✅ |
| Manifest | `populatePluginManifests(...)` | `populateServiceManifest(...)` | ✅ |
| Update stripper | `editStrippers.project[role].safeParse(data)` | Same logic in `buildProjectEditStrippers` | ✅ |

#### 4.8 Bulk Actions

| Aspect | Legacy | NestJS | Verdict |
|---|---|---|---|
| Parallel execution | `bulkExector` with `parallelBulkLimit` chunks | `Promise.allSettled` (no chunking) | ⚠️ Difference |
| Archive | Calls `archiveProject` (hooks + DB) | Calls `project.archive` (DB only, no hooks) | ⚠️ Behavioral difference |
| Lock/Unlock | `updateProject({ locked })` | `projectHooks.updateProjectLocked` | ✅ Equivalent |
| Replay | `replayHooks` | `projectHooks.replayHooks` | ✅ |

**⚠️ Behavioral difference**: Legacy bulk actions used chunked parallel execution (`parallelBulkLimit`). The NestJS version uses `Promise.allSettled` without chunking, which could overwhelm the database with a large number of concurrent operations.

---

### 5. Summary of Behavioral Differences

| Feature | Severity | Description |
|---|---|---|
| Hook failure handling on create | 🔴 High | Legacy fails fast (422), NestJS is fire-and-forget |
| Hook calls on archive | 🔴 High | Legacy calls `hook.project.delete`, NestJS has no hook call |
| Bulk action chunking | 🟡 Medium | Legacy chunks, NestJS does not |
| `includeUserType` in auth | 🔴 High | 2 tests failing — spec not updated |
| Missing `@StartActiveSpan` on update | 🟡 Medium | Telemetry gap |
| Controller specs missing | 🟡 Medium | 6 sub-module controllers untested |

### 6. Recommendations

1. **Fix `user.guard.spec.ts`** — Update `includeUserType: false` → `true` on lines 74 and 100.
2. **Add `@StartActiveSpan()` to `update()`** in `project.service.ts`.
3. **Clarify hook-to-event migration strategy** — If hooks are now handled by EventEmitter listeners, document this. If archive/delete hooks are not being emitted, add `project.delete` event emission to `project.service.ts:archive()`.
4. **Consider adding chunking to `ProjectBulkService`** for large-scale operations.
5. **Add controller specs** for the 6 sub-modules to match legacy `router.spec.ts` coverage.
6. **Consider structured logging** — Either inject pino or use `this.logger.log({ structuredObj }, 'message')` format for production observability parity.

---
*Reviewed by Hermes Agent — iso-behavior analysis vs legacy `apps/server`*

**Comment by shikanime** [comment]
## E2E Test Failure Root Cause Analysis

**Test:** `admin-stages.spec.ts:19:7 › should create a custom stage @e2e`
**Symptom:** After clicking "Create Project", URL stays at `/create-project` instead of redirecting to `/projects/:id`.

### Root Cause

The `POST /api/v1/projects` route now goes to **NestJS** (via nginx), but the **NestJS database is missing critical migrations** that exist in the legacy database:

| Migration | Present in Legacy? | Present in NestJS? |
|---|---|---|
| `20260204150335_add_system_roles` (admin permissions → 3) | ✅ | ❌ |
| `20260206105522_dso` (Administrateur Plateforme, permissions 3) | ✅ | ❌ |
| `20260226140001_add_tout_le_monde` (permissions 25608) | ✅ | ❌ |
| `20260415103000_set_root_admin_to_system_external` | ✅ | ❌ |

### Impact

The test user `tcolin` has:
- Keycloak group `/admin` → matches `AdminRole` with `permissions: 2` (MANAGE only) in NestJS DB
- `adminRoleIds: ['eadf604f-...']` → "Admin Locaux" with `permissions: 2` (MANAGE)
- **Missing:** "Tout le monde" role (`permissions: 25608`, includes `MANAGE_PROJECTS`)

Total admin permissions resolved by NestJS: `2n` (MANAGE only)
Required for `createProject`: `MANAGE_PROJECTS | MANAGE` = `10n`
Check: `2n & 10n = 0n` → **403 Forbidden**

The legacy server has "Tout le monde" role which grants `MANAGE_PROJECTS` (bit 3), so the test passes there.

### Fix

Ensure the NestJS Prisma database has all migrations applied up to and including `20260415103000_set_root_admin_to_system_external`. The NestJS `apps/server-nestjs/src/prisma/migrations/` directory is missing these migrations that exist in `apps/server/src/prisma/migrations/`.

Options:
1. **Copy missing migrations** from legacy to NestJS migrations directory, then run `prisma migrate deploy`
2. **Share the same DB** and ensure both apps point to the same PostgreSQL instance with all migrations applied
3. **Add a CI check** to verify NestJS DB schema is in sync with legacy

### Additional Note

The `user.guard.spec.ts` unit test failures (2 tests) are a separate issue — the spec asserts `includeUserType: false` but `UserGuard` now passes `true` (changed in commit `1c2b59eb3`, not updated in this PR).

File: apps/server-nestjs/src/modules/project-bulk/project-bulk.controller.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+@Controller('api/v1/projects-bulk')
+export class ProjectBulkController {
+  private readonly logger = new Logger(ProjectBulkController.name)
+
+  constructor(
+    @Inject(ProjectBulkService) private readonly projectBulk: ProjectBulkService,
+  ) {}
+
+  @Post('')
+  @HttpCode(202)
```
Comment (KepoParis) [inline apps/server-nestjs/src/modules/project-bulk/project-bulk.controller.ts]:
```suggestion
  @HttpCode(HttpStatus.OK)
```

**Review by KepoParis — CHANGES_REQUESTED** [review]

**Review by KepoParis — APPROVED** [review]
